templates/components/shared/llamaindex/embeddings.ts

```diff
@@ -0,0 +1,119 @@
+import fs from "fs";
+import {
+  BaseNode,
+  Document,
+  IngestionPipeline,
+  Metadata,
+  Settings,
+  SimpleNodeParser,
+  storageContextFromDefaults,
+  VectorStoreIndex,
+} from "llamaindex";
+import { DocxReader } from "llamaindex/readers/DocxReader";
+import { PDFReader } from "llamaindex/readers/PDFReader";
+import { TextFileReader } from "llamaindex/readers/TextFileReader";
+import crypto from "node:crypto";
+import { getDataSource } from "../engine";
+
+const MIME_TYPE_TO_EXT: Record<string, string> = {
+  "application/pdf": "pdf",
+  "text/plain": "txt",
+  "application/vnd.openxmlformats-officedocument.wordprocessingml.document":
+    "docx",
+};
+
+export async function uploadDocument(raw: string): Promise<string[]> {
+  const [header, content] = raw.split(",");
+  const mimeType = header.replace("data:", "").replace(";base64", "");
+  const fileBuffer = Buffer.from(content, "base64");
+  const documents = await loadDocuments(fileBuffer, mimeType);
+  const { filename, fileurl } = await saveDocument(fileBuffer, mimeType);
+
+  // update document metadata and mark documents to add to the vector store as private
+  for (const document of documents) {
+    document.metadata = {
+      ...document.metadata,
+      file_name: filename,
+      URL: fileurl,
```

Comment:
The file URL is for application serving specifically. I think we should not add that to the metadata.

Reply:
i think thuc did that because the FE is already using the URL for the public documents - so seems consistent with me

Reply:
but the marking is already done in runPipeline, keeps the code also cleaner to do it there

Reply:
we can construct the node url at response and it'd much better than fixing the url when embedding. Here is how i implemented in python:
https://github.com/run-llama/create-llama/pull/140/files#diff-4b95fdf4573bb6b3b4f07765683db5753ba993530e2de6346961790f239d1eaeR134

Reply:
if user change the server url then the constructed url for the old nodes will be broken, right?